Extend GraphPanel to Support multiple Graphs; Deprecate Derivation Feature Temporarily

FabianRolfMatthiasNoll opened the pull request:
This pull request enhances the GraphPanel by allowing it to support up to three graphs simultaneously. However, as a part of this update, the ability to perform derivation will be temporarily deprecated. The derivation feature will be reinstated in a future update, once the GraphPanel's integration with the new GUI is completed.

Details:

GraphPanel Enhancement: The GraphPanel has been upgraded to support the display of up to three graphs concurrently. This will allow users to visualize and analyze multiple graphs in parallel.

Temporary Deprecation of Derivation Feature: In light of the changes made to GraphPanel, the ability to derive functions will be temporarily unavailable. This feature will be re-enabled in an upcoming release, which will include the implementation of an enhanced Graphical User Interface (GUI) and a seamless interface between the GraphPanel and the GUI.

Continued Availability of Existing Features: The FunctionField, Calculate button, and Reset button will remain functional during this transition period. Users can continue to use these features for inputting functions and performing calculations until the new implementation is released.

Internal Handling of Polynomial Calculations: With this update, the handling of calculations specific to polynomials has been made more efficient by delegating this responsibility to the polynomial class itself. This refactoring improves the encapsulation and maintains cleaner separation of concerns.

These changes aim to set the stage for the upcoming enhancements to the GraphPanel, including a richer GUI and improved interaction capabilities.

FabianRolfMatthiasNoll committed:
Add Support to draw up to 3 Graphs

FabianRolfMatthiasNoll committed:
Add smoother calculation and adaptive tStep

FabianRolfMatthiasNoll committed:
cleanup

kantholz44mm commented on src/GraphPanel.java:
```diff
@@ -3,20 +3,19 @@
-    private double scale = 50.0;
-    private final Color backgroundColor = new Color(0.13f, 0.16f, 0.2f);
     private Vector2D lastMousePosition = null;
+    private double zoom = 1.0;
```
no need to have both zoom and scale, as they are directly proportional to one another and can be easily calculated (maybe implement a separate function, something like "getZoom" or "getScale" which does this on the fly). In general, we want as little state as necessary.

FabianRolfMatthiasNoll replied:
Done

kantholz44mm commented on src/GraphPanel.java:
```diff
@@ -3,20 +3,19 @@
 
 public class GraphPanel extends JPanel {
-    private double zoom = 1.0;
+    private final List<ParametricFunction> polynomials = new ArrayList<>();
```
The name of this variable suggests there are only polynomials, however these could be parametric, too.

FabianRolfMatthiasNoll replied:
Done

kantholz44mm commented on src/GraphPanel.java:
```diff
@@ -3,20 +3,19 @@
-    private List<Double> zeroPoints;
-    private List<Double> extremePoints;
+    private final Color backgroundColor = new Color(0.13f, 0.16f, 0.2f);
+    List<Color> colours = Arrays.asList(Color.WHITE, Color.BLUE, Color.GREEN, Color.RED, Color.PINK);
```
since the class is called "Color", this should be mirrorred in the variable names (it's 'correct' in the line above, for "backgroundColor"). Also, if "colours" is not modified anywhere, it should be final.

FabianRolfMatthiasNoll replied:
good catch completely missed that. Done

kantholz44mm commented on src/GraphPanel.java:
```diff
@@ -184,30 +173,42 @@ private void drawGrid(Graphics2D g2d, int width, int height, double step) {
         double maxT = toWorldCoordinates(new Vector2D(width, 0)).x;
 
-        int numSteps = 500;
+        int numSteps = (int) Math.max(5000, 10000 * zoom);
```
numSteps should not depend on the zoom level. For parametric functions, the user must provide the number of steps. For polynomials, the bounds for t are set dynamically by the left and right borders, anyway. The number of segments should only consider the screen width in this case. (x pixels/segment)

FabianRolfMatthiasNoll replied:
Done

kantholz44mm commented on src/GraphPanel.java:
```diff
@@ -184,30 +173,42 @@ private void drawGrid(Graphics2D g2d, int width, int height, double step) {
+                Vector2D p2 = toScreenCoordinates(polynomials.get(x).evaluate(t2));
+                Vector2D p3 = toScreenCoordinates(polynomials.get(x).evaluate(t3));
+
+                path.curveTo(p1.x, p1.y, p2.x, p2.y, p3.x, p3.y);
```
The function curveTo constructs a Bézier curve and takes as input the control points. The way p1-p3 are calculated, they are all on the same curve graph. This is very different from the way Bézier control points work. See [here](http://www.e-cartouche.ch/content_reg/cartouche/graphics/en/html/Curves_learningObject2.html) for example. This could be the cause of the recent 'jitteriness' of the graph with lower step counts. Changing this back to "lineTo" should fix that. This also allows us to calculate only 1/3 as many points, saving performance.

FabianRolfMatthiasNoll replied:
Done

FabianRolfMatthiasNoll committed:
Remove Scale

Scale and Zoom are proportional values. For that getScale was added to minimize states

FabianRolfMatthiasNoll committed:
Rename Array for functions

The functions in the array are not limited to be polynomials which resulted in bad naming

kantholz44mm commented on src/PolynomialFunction.java:
```diff
@@ -3,15 +3,18 @@
+    private double[] coefficients;
+    public List<Double> roots;
+    public List<Double> extremePoints;
     private static final Pattern TERM_PATTERN = Pattern.compile("([-+]?\\s*\\d*\\.?\\d*(?:/\\d+)*)?x(\\^(-?\\d+))?|([-+]?\\s*\\d+(/\\d+)?)");
```
again, these can be calculated and cached wherever needed. No need for state here, just have 'calcRoots' and 'getExtremePoints' return the lists directly. Also, if the TERM_PATTERN is only required in one place, it might be more sensible to declare it there.

FabianRolfMatthiasNoll replied:
The reason for that was that these values are only calculated once during creation of the polynomial instead of everytime the screen gets repainted => saving performance
But please correct me if i see something wrong here.
As for the TERM Pattern you are completely right. It was originally placed there for performance but i neglected the fact that the parsing only happens once for each function

kantholz44mm replied:
I think the GraphPanel may be the better location for this. Caching is a good idea, of course

FabianRolfMatthiasNoll replied:
Done

kantholz44mm commented on src/PolynomialFunction.java:
```diff
@@ -3,15 +3,18 @@
     private static final Pattern TERM_PATTERN = Pattern.compile("([-+]?\\s*\\d*\\.?\\d*(?:/\\d+)*)?x(\\^(-?\\d+))?|([-+]?\\s*\\d+(/\\d+)?)");
 
-    public PolynomialFunction(String polynomialString) {
+    public PolynomialFunction(String polynomialString, double start, double end) {
```
a PolynomialFunction should not have a 'start' and 'end'

FabianRolfMatthiasNoll replied:
So is the naming the problem? would something like leftBorder and rigthBorder  or similar be better?

kantholz44mm replied:
the function should be oblivious to its 'start' and 'end'. This also belongs in the GraphPanel imo.

FabianRolfMatthiasNoll replied:
Done

kantholz44mm commented on src/PolynomialFunction.java:
```diff
@@ -123,33 +126,31 @@ public void derive() {
 
-    public List<Double> getZeroPoints(double start, double end, double step) {
-        List<Double> zeros = new ArrayList<>();
+    private void calcRoots(double start, double end, double step) {
```
this should return the list of roots, directly.

FabianRolfMatthiasNoll replied:
Done

FabianRolfMatthiasNoll committed:
Fix naming typo

FabianRolfMatthiasNoll committed:
Revert function drawing and changed dependency of numSteps

FabianRolfMatthiasNoll committed:
Move calculation call to GraphPanel

The Polynomial doesn't have to know / shouldn't know about start/end etc. so the GraphPanel will call those functions after initialization. Also, it is directly used as a PolynomialFunction because it will never be a Parametric Function over this call.
